ocaml/xapi/xapi_pool_update.ml

```diff
@@ -98,8 +98,27 @@ let updates_to_attach_count_tbl : (string, int) Hashtbl.t = Hashtbl.create 10
 
 let updates_to_attach_count_tbl_mutex = Mutex.create ()
 
+let get_update_vbds ~__context ~vdi =
+  let dom0 = Helpers.get_domain_zero ~__context in
+  Db.VDI.get_VBDs ~__context ~self:vdi
+  |> List.filter (fun self -> Db.VBD.get_VM ~__context ~self = dom0)
+
+let assert_update_vbds_attached ~__context ~vdi =
+  let vbds = get_update_vbds ~__context ~vdi in
+  List.iter
+    (fun self ->
+      if not (Db.VBD.get_currently_attached ~__context ~self) then
+        let msg =
+          Printf.sprintf
+            "pool_update: expected VBD='%s' to be attached but it isn't!"
+            (Ref.string_of self)
+        in
+        raise Api_errors.(Server_error (internal_error, [msg])))
```

Comment:
A more informative message could be produced if all the unattached vbds are shown instead of just the first one.

This can be done with:
```ocaml
match List.filter (Fun.not vbd_attached) vbds with
| [] -> ()
| unattached_vbds -> raise ...
```

Reply:
I second this. I was about to suggest (but like Pau's code better):
I wonder whether it is cleaner to first find all non-attached VBDs (`List.filter`) and then raise an error for all of them. The current code bails out when the first one is found (which would be the typical case, I assume). It's still fine.

```
let dom0 = Helpers.get_domain_zero ~__context in
let missing = 
  Db.VDI.get_VBDs ~__context ~self:vdi
  |> List.filter (fun self -> Db.VBD.get_VM ~__context ~self = dom0)
  |> List.filter (fun self -> (Db.VBD.get_currently_attached ~__context ~self)
```